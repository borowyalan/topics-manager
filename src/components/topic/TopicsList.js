import React from 'react';
import Section from '../Section';
import Text from '../Text';
import Topic from './Topic';

const TopicsList = props => (
    <Section>
        <Text type="primary">Zg≈Çoszone tematy</Text>
        <Topic/>
    </Section>
);

export default TopicsList;
