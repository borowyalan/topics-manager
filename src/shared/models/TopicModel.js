import { UserModel } from './UserModel';
import { find } from 'lodash';

export default class TopicModel {
    title;
    description;
    likes;
    trainers;
    members;
    author; //GithubUserModel
    id;

    static fromBackendData(backendData) {
        const topic = Object.assign(new TopicModel(), backendData);
        topic.members = backendData.members
            ? backendData.members.map(member =>
                Object.assign(new UserModel(), member)
            )
            : [];
        topic.trainers = backendData.trainers
            ? backendData.trainers.map(trainer =>
                Object.assign(new UserModel(), trainer)
            )
            : [];
        return topic;
    }

    amIAttending(myself) {
        return myself && find(this.members, { email: myself.email });
    }

    amITrainer(myself) {
        return (
            myself &&
            (this.author.email === myself.email ||
                find(this.trainers, { email: myself.email }))
        );
    }
}
