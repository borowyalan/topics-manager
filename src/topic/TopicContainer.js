import React from 'react';
import PropTypes from 'prop-types';
import TopicsList from './components/TopicsList';
import { getTopics } from '../actions/topic.actions';
import { connect } from 'react-redux';
import TopicModel from '../shared/models/TopicModel';
import Container from '../shared/components/Container';


// tests for actions & thunk's dispatching
class TopicContainer extends React.Component {

    componentDidMount() {
        this.props.getTopics();
    }

    render() {
        const { topics, pending, error } = this.props;
        return (
            <Container type="section">
                {error && <p>{error}</p>}
                {!error && <TopicsList pending={pending} topics={topics}/>}
            </Container>
        );
    }
}

const mapStateToProps = ({ topic }) => ({
    topics: topic.topics,
    pending: topic.pending,
    error: topic.error
});

const mapDispatchToProps = dispatch => ({
    getTopics: () => dispatch(getTopics())
});

TopicContainer.propTypes = {
    getTopics: PropTypes.func.isRequired,
    topics: PropTypes.arrayOf(PropTypes.instanceOf(TopicModel)),
    pending: PropTypes.bool.isRequired,
    error: PropTypes.any
};

export default connect(mapStateToProps, mapDispatchToProps)(TopicContainer);
