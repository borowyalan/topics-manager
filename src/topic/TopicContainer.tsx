import React, { SFC } from 'react';
import { connect } from 'react-redux';
import Section from '../components/Section';
import { Topic } from '../shared/models/topic';
import TopicsList from './components/TopicsList';

export interface TopicContainerProps {
    topics: Topic[],
    pending: boolean,
    error?: any,
}

export const TopicContainer: SFC<TopicContainerProps> = ({ topics, pending, error }) => (
    <Section>
        {error && <p>{error}</p>}
        {!error && <TopicsList pending={pending} topics={topics}/>}
    </Section>
);

const mapStateToProps = ({ topic }) => ({
    topics: topic.topics,
    pending: topic.pending,
    error: topic.error,
});

export default connect(mapStateToProps)(TopicContainer);
