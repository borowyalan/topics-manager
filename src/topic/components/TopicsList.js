import React from 'react';
import Text from '../../components/Text';
import Topic from './Topic';
import PropTypes from 'prop-types';
import TopicModel from '../../shared/models/TopicModel';
import ActivityIndicator from '../../shared/components/ActivityIndicator';

const TopicsList = ({ topics, pending }) => (
    <React.Fragment>
        <Text type='primary'>Zg≈Çoszone tematy</Text>
        {pending && <ActivityIndicator/>}
        <ul>
            {topics.map((topic, index) => <Topic key={index} topic={topic}/>)}
        </ul>
    </React.Fragment>
);

TopicsList.propTypes = {
    topics: PropTypes.arrayOf(PropTypes.instanceOf(TopicModel)).isRequired,
    pending: PropTypes.bool
};

TopicsList.defaultProps = {
    topics: [],
    pending: false
};

export default TopicsList;
