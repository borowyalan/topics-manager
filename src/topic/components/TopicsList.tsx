import React, { SFC } from 'react';
import styled from 'styled-components';
import Text from '../../components/Text';
import ActivityIndicator from '../../shared/components/ActivityIndicator';
import { Topic as TopicInterface } from '../../shared/models/topic';
import Topic from './Topic';

export interface TopicListProps {
    topics: TopicInterface[],
    pending: boolean,
}

const HeadingText = styled(Text)`
    margin: 20px 0;
`;

const TopicsList: SFC<TopicListProps> = ({ topics = [], pending = false }) => (
    <React.Fragment>
        <HeadingText type="header">
            Zg≈Çoszone tematy
        </HeadingText>
        {pending && <ActivityIndicator/>}
        {topics.map((topic, index) => <Topic key={index} topic={topic}/>)}
    </React.Fragment>
);

export default TopicsList;
