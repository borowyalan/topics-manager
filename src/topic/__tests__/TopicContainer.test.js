import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import 'jest-styled-components';
import TopicContainer from '../TopicContainer';
import thunk from 'redux-thunk';
import TopicModel from '../../shared/models/TopicModel';
import TopicsList from '../components/TopicsList';

const mockedActionType = 'TopicContainerAction';

jest.mock('../../actions/topic.actions', () => ({
    getTopics: () => ({ type: mockedActionType }),
}));

configure({ adapter: new Adapter() });
const mockStore = configureMockStore([thunk]);

describe('<TopicContainer>', () => {
    const initialState = {
        auth: {
            loggedUser: false,
        },
        topic: {
            pending: true,
            error: 'test error message',
            topics: [TopicModel.fromBackendData({})],
        },
    };
    let wrapper;
    let store = mockStore(initialState);

    beforeEach(() => {
        store.clearActions();
        /**
         * help would be appreciated. we have to (?) dive, otherwise we will not have the real actions binded to
         * Component.
         * 1) Passing it as props is a fake test (i.e we don't know if the real one is called
         * 2) while using a dive, we have actions called immediately which may not be what we want
         *
         * I picked 2) as using 1) would only tell us that prop is called but we would lose the assurance that
         * Container is correctly 'connected' (i.e. has correctly mapped dispatchToProps
         *
         */
        wrapper = shallow(<TopicContainer store={store} />).dive();
    });

    it('should match snapshot with error', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should have topics array connected correctly from redux', () => {
        expect(wrapper.instance().props.topics).toBe(initialState.topic.topics);
    });

    it('should have pending state connected correctly from redux', () => {
        expect(wrapper.instance().props.pending).toBe(
            initialState.topic.pending
        );
    });

    it('should have error state connected correctly from redux', () => {
        expect(wrapper.instance().props.error).toBe(initialState.topic.error);
    });

    describe('when there are no errors', () => {
        const initialState = {
            auth: {
                loggedUser: false,
            },
            topic: {
                pending: true,
                error: null,
                topics: [
                    TopicModel.fromBackendData({}),
                    TopicModel.fromBackendData({}),
                ],
            },
        };
        let wrapper;
        let store = mockStore(initialState);

        beforeEach(() => {
            wrapper = shallow(<TopicContainer store={store} />).dive();
        });

        it('should match snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('should properly pass topics to rendered <TopicsList>', () => {
            //is a bit too complicated and should be splitted into context-based `describes` but for this one test it would be too much
            expect(
                wrapper
                    .find(TopicsList)
                    .at(0)
                    .props().topics.length
            ).toBe(2);
        });
    });
});
