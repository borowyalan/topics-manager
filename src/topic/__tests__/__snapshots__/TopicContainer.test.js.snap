// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopicContainer> should match snapshot with error 1`] = `
<Section>
  <p>
    test error message
  </p>
</Section>
`;

exports[`<TopicContainer> when there are no errors should match snapshot 1`] = `
<Section>
  <TopicsList
    pending={true}
    topics={
      Array [
        TopicModel {
          "members": Array [],
          "trainers": Array [],
        },
        TopicModel {
          "members": Array [],
          "trainers": Array [],
        },
      ]
    }
  />
</Section>
`;
